### PR TITLE
Balance refactor step one.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**
In the old implementation, we use leader count and capacity to calculate the score.
In balance loop, we select a highest score store and try to do balance including leader balance and follower balance.
It may have some problems, see #186 
To resolve this problem，we should
1）separate leader count and capacity balancer.
2）find a new algorithm to decide the priority of balance strategy dynamically.

This PR is only for 1) step, the balance strategy is leader balance first then capacity balance.
Although my test shows that the balance result is much better,  we should consider more and more rule conditions.

/cc @siddontang  @huachaohuang  @zhangjinpeng1987  @shenli

### CODE REVIEW

File: server/api/util.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/juju/errors"
+)
+
+func fromBody(r *http.Request, data interface{}) error {
+	body, err := ioutil.ReadAll(r.Body)
```
Comment (siddontang) [inline server/api/util.go]: should close body later

Comment (qiuyesuifeng) [inline server/api/util.go]: Yes, i missed it.

File: server/balancer_worker.go [hunk 2]
```diff
@@ -64,18 +66,20 @@ func (bw *balancerWorker) run() {
...
 		case <-bw.quit:
 			return
-		case <-ticker.C:
+		case <-timer.C:
 			err := bw.doBalance()
 			if err != nil {
 				log.Warnf("do balance failed - %v", errors.ErrorStack(err))
 			}
+
+			timer.Reset(time.Duration(bw.cfg.BalanceInterval) * time.Second)
```
Comment (siddontang) [inline server/balancer_worker.go]: I guess if we change config in another goroutine, here will raise race error.

File: server/cache.go [hunk 3]
```diff
@@ -552,28 +561,25 @@ func (s *storeInfo) usedRatio() float64 {
...
 type clusterInfo struct {
 	sync.RWMutex
 
-	meta        *metapb.Cluster
-	stores      map[uint64]*storeInfo
-	regions     *regionsInfo
-	clusterRoot string
+	meta          *metapb.Cluster
+	stores        map[uint64]*storeInfo
+	unknownStores map[uint64]struct{}
```
Comment (siddontang) [inline server/cache.go]: what is for unknown store?

Comment (qiuyesuifeng) [inline server/cache.go]: `unknown store` means we have not received the store status heartbeat.
So we cannot select this store as a candidate.

File: server/balancer_worker.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/balancer_worker.go]: That is a problem.
I leave a TODO in config.go, later we should check the config argument.

**Comment by siddontang** [comment]
LGTM

**Comment by huachaohuang** [comment]
LGTM

**Pull request merged**